### PR TITLE
Update artifact handling

### PR DESCRIPTION
**Pull request by joernott**
This pull request updates our usage of upload-artifact and download-artifact to v4. They introduced a breaking change which makes artifacts immutable, so we can't add more files to an already created artifact, a feature I relied on quite heavily. These are the changes:
- Actionlint complained about "number" in the inputs of dispatch_release.yml. I replaced that by string.
- I replaced v3 with v4 for all occurrences of upload-artifact and download-artifact. in a lot of actions
- Initially, we had an artifact named "logs" where all the logs were consolidated. In one of the later 3.x versions, I also added these logs to the artifacts which were created by every step. I now remove the functionality to add them to the consolidated "logs" archive. This will render the logfile.prefix setting obsolete.
- Initially, I created the "testplan" cached object to contain the initially generated testplan and then I used the begin-report action to add headers for the slack and summary report to the same archive. This has been streamlined by adding a new parameter "files" to the action and populating it with the files I initially uploaded. This eliminates one upload-artifact call as we add these files with begin-report now.
- The actions begin_report and append-report use unique artifact names starting with testplan- now. The action generate_report uses a new pattern matching feature to download all those artifacts matching testplan-* instead of the original "testplan" artifact. I am then creating a new "testplan" artifact from it afterwards and delete the testplan-* artifacts to clean up. The cleanup is still buggy, I have a ticket open with the author: https://github.com/GeekyEggo/delete-artifact/issues/22)
- **Breaking change**: Whenever we create a code coverage file for sonarcloud, we now need different names for every artifact. So, we need to define a different prefix for every phpunit and codeception run. That means in every test plan using sonarcloud, we need to set a unique value for coverage.prefix. To pick up all those coverage reports, the sonarcloud action now uses the __coverage_artifact__ parameter to define the pattern by which we choose the artifacts. In the universal workflow, I configured 'coverage-reports-**-${{ steps.sonarcloud_testplan_name.outputs.matrix_suffix }}', so all these coverage.prefix entries must start with "coverage-reports-".
- The output_artifact parameter of the son arcloud action was previously unused. I now use it to define the name for a consolidated artifact containing all reports. I also delete the artifacts which were consolidated.

### CODE REVIEW

**Review by snehasreeramini — CHANGES_REQUESTED** [review]

File: .github/workflows/call-universal_test_workflow.yml [hunk 1]
```diff
@@ -1389,22 +1383,14 @@ jobs:
...
           slack_webhook: ${{ secrets.SLACK_WEBHOOK_URL }}
           title: '${{ needs.init.outputs.finish_slack_title }}'
 
-      - name: Checkout the repo
-        uses: actions/checkout@v4
-
-      - name: 'clean_cache'
+      - name: 'Clean Cache'
         if: ${{ steps.generate_report.outputs.overall_status == 'success' }}
         uses: 'OXID-eSales/github-actions/clean_cache@v3'
```
Comment (snehasreeramini) [inline .github/workflows/call-universal_test_workflow.yml:1398]: here it is v3

File: sonarcloud/action.yml [hunk 2]
```diff
@@ -85,11 +86,22 @@ runs:
           sed -e 's|${{ inputs.strip_path }}||' -i ${FILE}
           COVERAGE_FILES="${COVERAGE_FILES},${FILE}"
           diff -y --color=always "${FILE}.orig" "${FILE}" || true
```
[the remark this answers is not in the document]
Comment (joernott) [reply sonarcloud/action.yml:88]: This helps with debugging. As we have to consolidate all those coverage-reports into one single artifact, we can have the modified versions alongside the unmodified files in that artifact.
Before, the coverage-reports artifact only contained the unmodified files as it was created by the test jobs before and we only unpacked it here. Now, that we have to create a single artifact and delete all those single files created by the previous jobs, we can as well add originals and patched files together and if someone wonders, why files are not covered, he can see what the pattern replacement did.

File: phpcs/action.yml [hunk 3]
```diff
@@ -102,7 +102,7 @@ runs:
 
     - name: Upload Log
       if: ${{ always() && steps.diff.outputs.skip == 'false' }}
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@v4
       with:
         name: ${{ inputs.logfile_artifact }}
```
Comment (snehasreeramini) [inline phpcs/action.yml:107]: this logfile_artifact is removed right

File: tests/github_actions/defaults/defaults.yml [hunk 4]
```diff
@@ -669,7 +681,7 @@ sonarcloud:
   custom_script_container: ''
 
   coverage:
-    prefix: *coverage_prefix
+    prefix: coverage-report
```
Comment (snehasreeramini) [inline tests/github_actions/defaults/defaults.yml:684]: isin't coverage-reports-sonarcloud

File: begin_report/action.yml [hunk 5]
```diff
@@ -41,11 +50,12 @@ runs:
 
     - name: Upload Artifacts
       if: always()
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@v4
       with:
-        name: testplan
+        name: testplan-000_header
```
[the remark this answers is not in the document]
Comment (joernott) [reply begin_report/action.yml:55]: The reason for this whole change is that we can't have a single "testplan" artifact which is created once and then every job adds stuff to it. As I wrote in the explanation above, these artifacts are now immutable. They can't be changed any more. So we create a lot of small artifacts which all start with testplan- instead. In the generate-report step, we unpack all those testplan-* artifacts, generate our reports, pack everything together in the final "testplan" artifact and delete all the testplan-* artifacts as their content is now in the single testplan artifact.

File: tests/github_actions/defaults/defaults.yml [hunk 6]
```diff
@@ -622,7 +634,7 @@ runtest:
     prefix: 'deprecated_tests_artifacts'
   coverage:
     path: 'source/deprecated_tests_coverage.xml'
-    prefix: *coverage_prefix
+    prefix: coverage-report-runtest
```
Comment (snehasreeramini) [inline tests/github_actions/defaults/defaults.yml:637]: coverage-reports-runtest

**Review by snehasreeramini — CHANGES_REQUESTED** [review]

**Review by snehasreeramini — APPROVED** [review]

**Pull request merged**